### PR TITLE
Updates (npm+gulp, docker, anchorjs, dev mode, composer packages)

### PR DESCRIPTION
**Pull request by webimpress**
- added development mode
- added **docker** configuration
- updated all packages
- added plugin to render tables in documentation - https://apigility.org/documentation/auth/advanced
- replaced **bower** with **npm+gulp**;
  all bower components are now removed, assets are compiled on build process (`composer build`)
- added **`anchorjs`** - links to proper section in documentation,
- fixed video layout and served video directly from youtube, without using any other players: https://apigility.org/video
- lossless image optimization
- cleared `.gitignore` from IDE and OS-specific paths
- enabled app config cache (cache is cleared on building project - `composer build`)
- updated links to https and year to 2017 (or `date('Y')` in layout)
- updated copyrights: "by Zend, a Rouge Wave Company"
- added contact icons in footer

screen shot 2017-09-26 at 09 38 24

### CODE REVIEW

**Comment by weierophinney** [comment]
The jump table preview is too big to display... so I have one HUGE question: did you also update the scripts under `.aws/`, which are used by AWS CodeDeploy to deploy the website?

If not, I need detailed instructions on:

- what dependencies are required to build the site
- the precise steps for building the site

otherwise, merging this right now is a complete shot in the dark for us, and can lead to a broken web site.

Thanks!

**Comment by webimpress** [comment]
@weierophinney As we discussed on Slack I've updated `.aws/before_install.sh` to install needed dependencies - also with nginx and php7.0 + modules. Please see commit d2d91c6.

I've tested it with `ubuntu 16.04` and it works fine. I believe it should be OK.
Do you have any possibility to test it on AWS before merging?
Is merge going to update AWS automatically?

File: config/development.config.php [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<?php
```
Comment (weierophinney) [inline config/development.config.php]: Shouldn't this file be excluded via `.gitignore`? It will be copied from `development.config.php.dist` when in development mode...

File: module/Application/view/application/video/index.phtml [hunk 2]
```diff
@@ -8,10 +8,10 @@
...
 
     <div class="grid" style="padding-left: 10px; margin-top: 20px;">
-        <div class="videoContent">
-            <video id="agPlayer" controls preload="none" width="960" height="540" style="width: 100%; height: 100%;">
-                <source src="http://www.youtube.com/watch?v=ZBSX1AB6LcM" type="video/youtube">
-            </video>
+        <div class="media-wrapper">
+            <iframe width="960" height="540"
+                    src="https://www.youtube.com/embed/ZBSX1AB6LcM"
+                    frameborder="0" allowfullscreen></iframe>
```
Comment (weierophinney) [inline module/Application/view/application/video/index.phtml]: I don't understand these changes; why are you reverting to an iframe instead of using the HTML5 video tag?

File: module/Application/view/application/video/template.phtml [hunk 3]
```diff
@@ -8,10 +8,10 @@
...
 
     <div class="grid" style="padding-left: 10px; margin-top: 20px;">
-        <div class="videoContent">
-            <video id="agPlayer" controls preload="none" width="960" height="540" style="width: 100%; height: 100%;">
-                <source src="http://www.youtube.com/watch?v=<?= $this->mainVideo['id'] ?>" type="video/youtube">
-            </video>
+        <div class="media-wrapper">
+            <iframe width="960" height="540"
+                    src="https://www.youtube.com/embed/<?= $this->mainVideo['id'] ?>"
+                    frameborder="0" allowfullscreen></iframe>
```
Comment (weierophinney) [inline module/Application/view/application/video/template.phtml]: Same comment here about video tag...

File: module/Application/view/layout/layout.phtml [hunk 4]
```diff
@@ -13,17 +13,15 @@
                 'type' => 'image/vnd.microsoft.icon',
                 'href' => $this->basePath() . '/img/favicon.ico'
             ])
-            ->appendStylesheet($this->basePath() . '/bower_components/mediaelement/build/mediaelementplayer.min.css')
-            ->appendStylesheet($this->basePath() . '/css/font.css')
-            ->appendStylesheet($this->basePath() . '/css/screen.css') ?>
+            ->appendStylesheet($this->basePath() . '/' . $this->asset('css/styles.css'))
+            ->appendStylesheet('https://cdn.jsdelivr.net/npm/font-awesome@4.7.0/css/font-awesome.min.css')
```
Comment (weierophinney) [inline module/Application/view/layout/layout.phtml]: Why is this one via a CDN, and not part of the styles we serve? If we serve it, no network connection is required when testing...

File: module/Application/view/layout/layout.phtml [hunk 5]
```diff
@@ -58,17 +56,26 @@
 
     <footer>
         <div class="grid">
-            <div class="unit one-third center-on-mobiles">
+            <div class="unit half center-on-mobiles">
```
Comment (weierophinney) [inline module/Application/view/layout/layout.phtml]: Why are you changing from 1:2 to 1:1 ratios for the columns?

File: phpcs.xml [hunk 6]
```diff
@@ -1,35 +1,9 @@
...
-        </properties>
-    </rule>
-    <rule ref="PSR1.Files.SideEffects">
-        <exclude-pattern>public/index.php</exclude-pattern>
-        <exclude-pattern>bin/install.php</exclude-pattern>
-    </rule>
-    <rule ref="Generic.Files.LineLength.TooLong">
-        <exclude-pattern>config/autoload/global.php</exclude-pattern>
-        <exclude-pattern>bin/install.php</exclude-pattern>
-    </rule>
```
Comment (weierophinney) [inline phpcs.xml]: A number of these rules provided exclusions for configuration and script files; why were those removed?

File: config/development.config.php (hunk 1 above)
Comment (webimpress) [reply config/development.config.php]: Good point, fixed in 8c5058535c5d10ff9fea2050a18c59e6f8106f66

File: module/Application/view/application/video/index.phtml (hunk 2 above)
Comment (webimpress) [reply module/Application/view/application/video/index.phtml]: With video tag it's not gonna work on all browsers. Then we have to use some JS libraries to add support - previously we used.... hm.. mediaelement (?) Now 'responsibility' to display the video properly we moved on YouTube and I think it's better, as we are using videos only from YT.

File: module/Application/view/layout/layout.phtml (hunk 4 above)
Comment (webimpress) [reply module/Application/view/layout/layout.phtml]: The same we are doing in zf3-web: https://github.com/zendframework/zf3-web/blob/master/templates/layout/default.phtml#L11

I agree that on testing it requires network connection, but I don't think so it's a big issue. When we compile it into our assets then we will serve much bigger css file. I think external library like FA should be server from their CDN, and it could be already cached in many users browser.

File: module/Application/view/layout/layout.phtml (hunk 5 above)
Comment (webimpress) [reply module/Application/view/layout/layout.phtml]: It's in footer, we had before two columns with width: 'one-third' + 'half', now we have both 'half'. We need wider column there, because of icons I've added in footer - please see image attached in the description of the PR.

File: phpcs.xml (hunk 6 above)
Comment (webimpress) [reply phpcs.xml]: I've done it because of moving to zend-coding-standard. I've reverted the exclusion in 927fc260a2a4558c690f1be352eddfd0f8fd0260
We are not running anywhere cs-check - there is no any Travis CI config with the repository...

File: module/Application/view/application/video/template.phtml (hunk 3 above)
Comment (webimpress) [reply module/Application/view/application/video/template.phtml]: Please see comment above https://github.com/zfcampus/apigility.org/pull/23#discussion_r148141084

**Comment by webimpress** [comment]
@weierophinney I've replied on your comments and pushed changes as noted.

Images (as noted in the PR description) are _not changed_ in visible way. I've applied only 'lossless image optimization' so pixel-to-pixel are exactly the same but these takes less space.

As I said, I've applied only lossless optimization, google suggest even more:
https://developers.google.com/speed/docs/insights/OptimizeImages
but it changes slightly images. See also: https://github.com/zendframework/zf3-web/pull/79

File: module/Application/view/layout/layout.phtml (hunk 5 above)
Comment (weierophinney) [reply module/Application/view/layout/layout.phtml]: Previously, we had two columns, one as one-third and one as two-thirds; the next diffed section demonstrates the change from two-thirds to half!

However, if we need a wider column, that's enough justification.

**Comment by weierophinney** [comment]
> Do you have any possibility to test it on AWS before merging?
> Is merge going to update AWS automatically?

We don't really have an ability to test on AWS first, as we need it to run through our CodeDeploy setup, which is only for production.

Merging and pushing to master _does_ automatically trigger the CodeDeploy listener. However, we have the ability to roll back to previous images with CodeDeploy as well, so if this goes wrong, we can still go back.

File: module/Application/view/layout/layout.phtml (hunk 5 above)
Comment (webimpress) [reply module/Application/view/layout/layout.phtml]: Oh, yeah, you have right, sorry. We didn't use the whole space in right column so it's not breaking anything. Now both have the same width: 50-50. Tested ;-)

**Comment by weierophinney** [comment]
Okay, deployment failed. I'm rolling back, and then will look through logs to determine what happened.

**Comment by webimpress** [comment]
@weierophinney Yeah, just seen error 500... I wonder what's wrong.

**Comment by weierophinney** [comment]
Error was missing php7.0-curl, which caused the script immediately prior to building assets to fail. Resolved now!
